cpp/dolfinx/fem/dofmapbuilder.cpp

```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
+
+  for (std::size_t i = 0; i < num_cell_types; ++i)
+  {
+    mesh::CellType cell_type = topology.entity_types(D)[i];
+    const std::vector<std::vector<std::vector<int>>> entity_dofs
+        = element_dof_layouts[i].entity_dofs_all();
+    assert(entity_dofs.size() == D + 1);
+
+    // Loop over cells of this type, and build dofmap
+    std::int32_t num_cells = topology.index_maps(D)[i]->size_local()
+                             + topology.index_maps(D)[i]->num_ghosts();
+    std::int32_t dofmap_width = element_dof_layouts[i].num_dofs();
+    dofs[i].width = dofmap_width;
+    dofs[i].array.resize(num_cells * dofmap_width);
+    LOG(INFO) << "Cell type:" << i << ", dofmap:" << num_cells << "x"
+              << dofmap_width;
+
+    std::int32_t dofmap_offset = 0;
+    for (std::int32_t c = 0; c < num_cells; ++c)
     {
-      if (needs_entities[d])
+      // Wrap dofs for cell c
+      std::span<std::int32_t> dofs_c(dofs[i].array.data() + dofmap_offset,
+                                     dofmap_width);
+      dofmap_offset += dofmap_width;
+
+      // Iterate over required entities for this element, dimension and type
+      for (std::size_t k = 0; k < required_dim_et.size(); ++k)
       {
+        std::size_t d = required_dim_et[k].first;
+        std::size_t et = required_dim_et[k].second;
+        mesh::CellType e_type = topology.entity_types(d)[et];
+
         const std::vector<std::vector<int>>& e_dofs_d = entity_dofs[d];
 
-        // Iterate over each entity of current dimension d
-        std::size_t num_entity_dofs = e_dofs_d[0].size();
+        // Iterate over each entity of current dimension d and type et
         const std::int32_t* c_to_e
-            = (d == D) ? nullptr : connectivity[d]->links(c).data();
+            = ((std::size_t)d == D)
+                  ? nullptr
+                  : topology.connectivity({D, i}, {d, et})->links(c).data();
+
+        int w = 0;
         for (std::size_t e = 0; e < e_dofs_d.size(); ++e)
         {
-          assert(e_dofs_d[e].size() == num_entity_dofs);
-          std::int32_t e_index_local = (d == D) ? c : c_to_e[e];
-
-          // Loop over dofs belonging to entity e of dimension d (d, e)
-          // d: topological dimension
-          // e: local entity index
-          // dof_local: local index of dof at (d, e)
-          for (std::size_t i = 0; i < num_entity_dofs; ++i)
+          // Skip entities of wrong type (e.g. for facets of prism)
+          // Use separate connectivity index 'w' which only advances for
+          // correct entities
+          if (mesh::cell_entity_type(cell_type, d, e) == e_type)
           {
-            int dof_local = e_dofs_d[e][i];
-            dofs_c[dof_local]
-                = offset_local + num_entity_dofs * e_index_local + i;
+            std::size_t num_entity_dofs = e_dofs_d[e].size();
+            assert((int)num_entity_dofs == num_entity_dofs_et[k]);
+            std::int32_t e_index_local = ((std::size_t)d == D) ? c : c_to_e[w];
```

Comment:
```suggestion
            std::int32_t e_index_local = (std::size_t)d == D ? c : c_to_e[w];
```
Get rid of the cast?